test/e2e/federated-service.go

```diff
@@ -0,0 +1,209 @@
+
+	BeforeEach(func() {
+		contexts := f.GetUnderlyingFederatedContexts()
+
+		for _, context := range contexts {
+			framework.Logf("Creating cluster object: %s (%s)", context.Name, context.Cluster.Cluster.Server)
+			cluster := federation.Cluster{
+				ObjectMeta: api.ObjectMeta{
+					Name: context.Name,
+				},
+				Spec: federation.ClusterSpec{
+					ServerAddressByClientCIDRs: []federation.ServerAddressByClientCIDR{
+						{
+							ClientCIDR:    "0.0.0.0/0",
+							ServerAddress: context.Cluster.Cluster.Server,
+						},
+					},
+				},
+			}
+			_, err := f.FederationClientset.Federation().Clusters().Create(&cluster)
+			framework.ExpectNoError(err, "Creating cluster")
+		}
+
+		By("Obtaining a list of all the clusters")
+		clusterList, err := f.FederationClientset.Federation().Clusters().List(api.ListOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		framework.Logf("%d clusters found", len(clusterList.Items))
+
+		for _, cluster := range clusterList.Items {
+			framework.Logf("Creating a clientset for the cluster %s", cluster.Name)
+
+			Expect(framework.TestContext.KubeConfig).ToNot(Equal(""), "KubeConfig must be specified to load clusters' client config")
+			kubecfg, err := clientcmd.LoadFromFile(framework.TestContext.KubeConfig)
+			framework.ExpectNoError(err, "error loading KubeConfig: %v", err)
+
+			cfgOverride := &clientcmd.ConfigOverrides{
+				ClusterInfo: clientcmdapi.Cluster{
+					Server: cluster.Spec.ServerAddressByClientCIDRs[0].ServerAddress,
+				},
+			}
+			ccfg := clientcmd.NewNonInteractiveClientConfig(*kubecfg, cluster.Name, cfgOverride, clientcmd.NewDefaultClientConfigLoadingRules())
+			cfg, err := ccfg.ClientConfig()
+			Expect(err).NotTo(HaveOccurred())
+
+			cfg.QPS = KubeAPIQPS
+			cfg.Burst = KubeAPIBurst
+			clset := release_1_3.NewForConfigOrDie(restclient.AddUserAgent(cfg, UserAgentName))
+			clusterClientSets = append(clusterClientSets, clset)
+		}
+	})
+
+	It("should be able to discover a federated service", func() {
+		framework.SkipUnlessFederated(f.Client)
+
+		createService(f.FederationClientset, clusterClientSets, f.Namespace.Name)
+
+		discoverService(f)
+	})
+
+	It("should be able to discover a non-local federated service", func() {
+		framework.SkipUnlessFederated(f.Client)
+
+		createService(f.FederationClientset, clusterClientSets, f.Namespace.Name)
+
+		// Delete a federated service shard in the default e2e Kubernetes cluster.
+		err := f.Clientset_1_3.Core().Services(f.Namespace.Name).Delete(FederatedServiceName, &api.DeleteOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		waitForFederatedServiceShard(f.Clientset_1_3, f.Namespace.Name, nil, 0)
+
+		discoverService(f)
+	})
+})
+
+func waitForFederatedServiceShard(cs *release_1_3.Clientset, namespace string, service *api.Service, numSvcs int) {
+	By("Fetching a federated service shard")
+	var clSvcList *v1.ServiceList
+	for t := time.Now(); time.Since(t) < FederatedServiceTimeout; time.Sleep(framework.Poll) {
```

Comment:
nit: There are some library functions (wait*) in our test libraries that will do this for you.  Better to use the standard libraries than roll your own retry/wait logic here.